Remove flakiness of TestServiceRepairResumeAllRangesIntegration for Tablets

TestServiceRepairResumeAllRangesIntegration might fail when executed against tablet env:
```
2024-05-02T15:32:43.4828588Z     service_repair_integration_test.go:1095: Overall redundant ranges 39
2024-05-02T15:32:43.4829937Z     service_repair_integration_test.go:1097: Expected less redundant token ranges than 39
2024-05-02T15:32:43.4831372Z --- FAIL: TestServiceRepairResumeAllRangesIntegration (91.77s)
```

As tablet tables are always repaired from scratch when repair is restarted (5 repairs * 8 tablets in table ~ 40 re-repaired ranges).
The test should be changed so that it takes tablets into consideration.

--- a/pkg/service/repair/service_repair_integration_test.go
+++ b/pkg/service/repair/service_repair_integration_test.go
@@ -1055,6 +1055,45 @@ func TestServiceRepairResumeAllRangesIntegration(t *testing.T) {
 		t.Fatal("Repair failed without error")
 	}
 
+	validate := func(tab string, tr []scyllaclient.TokenRange) (redundant int, err error) {
+		sort.Slice(tr, func(i, j int) bool {
+			return tr[i].StartToken < tr[j].StartToken
+		})
+
+		// Check that full token range is repaired
+		if tr[0].StartToken != dht.Murmur3MinToken {
+			return 0, errors.Errorf("expected min token %d, got %d", dht.Murmur3MinToken, tr[0].StartToken)
+		}
+		if tr[len(tr)-1].EndToken != dht.Murmur3MaxToken {
+			return 0, errors.Errorf("expected max token %d, got %d", dht.Murmur3MaxToken, tr[len(tr)-1].EndToken)
+		}
+
+		// Check if repaired token ranges are continuous
+		for i := 1; i < len(tr); i++ {
+			// Redundant ranges might occur due to pausing repair,
+			// but there shouldn't be much of them.
+			if tr[i-1] == tr[i] {
+				redundant++
+				continue
+			}
+			if tr[i-1].EndToken != tr[i].StartToken {
+				return 0, errors.Errorf("non continuous token ranges %v and %v", tr[i-1], tr[i])
+			}
+		}
+
+		return redundant, nil
+	}
+
+	// Tablet tables don't support resuming repair, so in order to check if repair
+	// started from scratch, we need to remove all repaired ranges up to this point.
+	ringDescriber := scyllaclient.NewRingDescriber(ctx, h.Client)
+	for tab, _ := range doneRanges {
+		_, err := validate(tab, doneRanges[tab])
+		if err != nil && ringDescriber.IsTabletKeyspace(strings.Split(tab, ".")[0]) {
+			doneRanges[tab] = nil
+		}
+	}
+
 	Print("When: run fifth repair till it finishes")
 	h.RunID = uuid.NewTime()
 	if err := h.runRegularRepair(ctx, props); err != nil {
@@ -1066,31 +1105,11 @@ func TestServiceRepairResumeAllRangesIntegration(t *testing.T) {
 	for tab, tr := range doneRanges {
 		t.Logf("Checking table %s", tab)
 
-		sort.Slice(tr, func(i, j int) bool {
-			return tr[i].StartToken < tr[j].StartToken
-		})
-
-		// Check that full token range is repaired
-		if tr[0].StartToken != dht.Murmur3MinToken {
-			t.Fatalf("Expected min token %d, got %d", dht.Murmur3MinToken, tr[0].StartToken)
+		r, err := validate(tab, tr)
+		if err != nil {
+			t.Fatal(err)
 		}
-		if tr[len(tr)-1].EndToken != dht.Murmur3MaxToken {
-			t.Fatalf("Expected max token %d, got %d", dht.Murmur3MaxToken, tr[len(tr)-1].EndToken)
-		}
-
-		// Check if repaired token ranges are continuous
-		for i := 1; i < len(tr); i++ {
-			// Redundant ranges might occur due to pausing repair,
-			// but there shouldn't be much of them.
-			if tr[i-1] == tr[i] {
-				t.Logf("Redundant range %v", tr[i])
-				redundant++
-				continue
-			}
-			if tr[i-1].EndToken != tr[i].StartToken {
-				t.Fatalf("Non continuous token ranges %v and %v", tr[i-1], tr[i])
-			}
-		}
+		redundant += r
 	}
 
 	t.Logf("Overall redundant ranges %d", redundant)

--- a/pkg/service/restore/service_restore_integration_test.go
+++ b/pkg/service/restore/service_restore_integration_test.go
@@ -224,11 +224,20 @@ func TestRestoreGetTargetUnitsViewsIntegration(t *testing.T) {
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
+			h := h
+			h.T = t
 			b, err := os.ReadFile(tc.input)
 			if err != nil {
 				t.Fatal(err)
 			}
 			b = jsonutil.Set(b, "snapshot_tag", tag)
+
+			var target Target
+			if err = json.Unmarshal(b, &target); err != nil {
+				t.Fatal(err)
+			}
+			h.shouldSkipTest(target)
+
 			target, units, views, err := h.service.GetTargetUnitsViews(ctx, h.ClusterID, b)
 			if err != nil {
 				t.Fatal(err)
@@ -577,6 +586,8 @@ func smokeRestore(t *testing.T, target Target, keyspace string, loadCnt, loadSiz
 		srcSession   = CreateSessionAndDropAllKeyspaces(t, srcH.Client)
 	)
 
+	dstH.shouldSkipTest(target)
+
 	// Restore should be performed on user with limited permissions
 	if err := createUser(dstSession, user, "pass"); err != nil {
 		t.Fatal(err)
@@ -643,6 +654,8 @@ func restoreWithAgentRestart(t *testing.T, target Target, keyspace string, loadC
 		srcSession   = CreateSessionAndDropAllKeyspaces(t, srcH.Client)
 		ctx          = context.Background()
 	)
+
+	dstH.shouldSkipTest(target)
 
 	// Restore should be performed on user with limited permissions
 	if err := createUser(dstSession, user, "pass"); err != nil {
@@ -747,6 +760,8 @@ func restoreWithResume(t *testing.T, target Target, keyspace string, loadCnt, lo
 		ctx1, cancel1 = context.WithCancel(context.Background())
 		ctx2, cancel2 = context.WithCancel(context.Background())
 	)
+
+	dstH.shouldSkipTest(target)
 
 	// Restore should be performed on user with limited permissions
 	if err := createUser(dstSession, user, "pass"); err != nil {
@@ -913,6 +928,8 @@ func restoreWithVersions(t *testing.T, target Target, keyspace string, loadCnt, 
 		srcSession   = CreateSessionAndDropAllKeyspaces(t, srcH.Client)
 		ctx          = context.Background()
 	)
+
+	dstH.shouldSkipTest(target)
 
 	status, err := srcH.Client.Status(ctx)
 	if err != nil {
@@ -1118,6 +1135,8 @@ func restoreViewCQLSchema(t *testing.T, target Target, keyspace string, loadCnt,
 		srcSession   = CreateSessionAndDropAllKeyspaces(t, srcH.Client)
 	)
 
+	dstH.shouldSkipTest(target)
+
 	Print("When: Create Restore user")
 	if err := createUser(dstSession, user, "pass"); err != nil {
 		t.Fatal(err)
@@ -1205,6 +1224,8 @@ func restoreViewSSTableSchema(t *testing.T, schemaTarget, tablesTarget Target, k
 		dstSession   = CreateSessionAndDropAllKeyspaces(t, dstH.Client)
 		srcSession   = CreateSessionAndDropAllKeyspaces(t, srcH.Client)
 	)
+
+	dstH.shouldSkipTest(schemaTarget, tablesTarget)
 
 	Print("When: Create Restore user")
 	if err := createUser(dstSession, user, "pass"); err != nil {
@@ -1302,6 +1323,8 @@ func restoreAllTables(t *testing.T, schemaTarget, tablesTarget Target, keyspace 
 		dstSession   = CreateSessionAndDropAllKeyspaces(t, dstH.Client)
 		srcSession   = CreateSessionAndDropAllKeyspaces(t, srcH.Client)
 	)
+
+	dstH.shouldSkipTest(schemaTarget, tablesTarget)
 
 	// Ensure clean scylla tables
 	if err := cleanScyllaTables(srcSession); err != nil {
@@ -1409,6 +1432,8 @@ func restoreAlternator(t *testing.T, schemaTarget, tablesTarget Target, keyspace
 		dstSession   = CreateSessionAndDropAllKeyspaces(t, dstH.Client)
 		srcSession   = CreateSessionAndDropAllKeyspaces(t, srcH.Client)
 	)
+
+	dstH.shouldSkipTest(schemaTarget, tablesTarget)
 
 	// Restore should be performed on user with limited permissions
 	if err := createUser(dstSession, user, "pass"); err != nil {
@@ -1879,4 +1904,17 @@ func tombstoneGCMode(s gocqlx.Session, keyspace, table string) (string, error) {
 		}
 	}
 	return "", errors.New("unknown mode")
+}
+
+func (h *restoreTestHelper) shouldSkipTest(targets ...Target) {
+	for _, target := range targets {
+		if target.RestoreSchema {
+			if err := IsRestoreSchemaSupported(context.Background(), h.Client); err != nil {
+				if errors.Is(err, ErrRestoreSchemaUnsupportedScyllaVersion) {
+					h.T.Skip(err)
+				}
+				h.T.Fatal(err)
+			}
+		}
+	}
 }

--- a/pkg/service/restore/worker.go
+++ b/pkg/service/restore/worker.go
@@ -110,7 +110,7 @@ func (w *worker) initTarget(ctx context.Context, properties json.RawMessage) err
 	}
 
 	if t.RestoreSchema {
-		if err := isRestoreSchemaSupported(ctx, w.client); err != nil {
+		if err := IsRestoreSchemaSupported(ctx, w.client); err != nil {
 			return err
 		}
 	}
@@ -164,9 +164,14 @@ func (w *worker) initTarget(ctx context.Context, properties json.RawMessage) err
 	return nil
 }
 
+// ErrRestoreSchemaUnsupportedScyllaVersion means that restore schema procedure is not safe for used Scylla configuration.
+var ErrRestoreSchemaUnsupportedScyllaVersion = errors.Errorf("restore into cluster with given ScyllaDB version and consistent_cluster_management is not supported. " +
+	"See https://manager.docs.scylladb.com/stable/restore/restore-schema.html for a workaround.")
+
+// IsRestoreSchemaSupported checks if restore schema procedure is supported for used Scylla configuration.
 // Because of #3662, there is no way fo SM to safely restore schema into cluster with consistent_cluster_management
 // and version higher or equal to OSS 5.4 or ENT 2024. There is a documented workaround in SM docs.
-func isRestoreSchemaSupported(ctx context.Context, client *scyllaclient.Client) error {
+func IsRestoreSchemaSupported(ctx context.Context, client *scyllaclient.Client) error {
 	const (
 		DangerousConstraintOSS = ">= 6.0, < 2000"
 		DangerousConstraintENT = ">= 2024.2, > 1000"
@@ -214,8 +219,7 @@ func isRestoreSchemaSupported(ctx context.Context, client *scyllaclient.Client) 
 	}
 
 	if raftSchema && !raftIsSafe {
-		return errors.Errorf("restore into cluster with given ScyllaDB version and consistent_cluster_management is not supported. " +
-			"See https://manager.docs.scylladb.com/stable/restore/restore-schema.html for a workaround.")
+		return ErrRestoreSchemaUnsupportedScyllaVersion
 	}
 	return nil
 }

